store/localcachelayer/terms_of_service_layer.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package localcachelayer
+
+import (
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/store"
+)
+
+type LocalCacheTermsOfServiceStore struct {
+	store.TermsOfServiceStore
+	rootStore *LocalCacheStore
+}
+
+const (
+	termsOfServiceCacheName = "TermsOfServiceStore"
+)
+
+func (s *LocalCacheTermsOfServiceStore) handleClusterInvalidateTermsOfService(msg *model.ClusterMessage) {
+	if msg.Data == CLEAR_CACHE_MESSAGE_DATA {
+		s.rootStore.termsOfServiceCache.Purge()
+	} else {
+		s.rootStore.termsOfServiceCache.Remove(msg.Data)
+	}
+}
+
+func (s LocalCacheTermsOfServiceStore) ClearCaches() {
+	s.rootStore.doClearCacheCluster(s.rootStore.termsOfServiceCache)
+
+	if s.rootStore.metrics != nil {
+		s.rootStore.metrics.IncrementMemCacheInvalidationCounter("Terms Of Service - Purge")
+	}
+}
+
+func (s LocalCacheTermsOfServiceStore) InvalidateTermsOfService(termsOfServiceId string) {
+	s.rootStore.doInvalidateCacheCluster(s.rootStore.termsOfServiceCache, termsOfServiceId)
+	if s.rootStore.metrics != nil {
+		s.rootStore.metrics.IncrementMemCacheInvalidationCounter("Terms Of Service - Remove by TermsOfServiceId")
+	}
+}
+
+func (s LocalCacheTermsOfServiceStore) Save(termsOfService *model.TermsOfService) (*model.TermsOfService, *model.AppError) {
+	tos, err := s.TermsOfServiceStore.Save(termsOfService)
+	if err != nil {
+		return nil, err
+	}
+	s.rootStore.termsOfServiceCache.AddWithDefaultExpires(termsOfService.Id, termsOfService)
+	return tos, nil
+}
+
+func (s LocalCacheTermsOfServiceStore) GetLatest(allowFromCache bool) (*model.TermsOfService, *model.AppError) {
+	if allowFromCache {
+		if s.rootStore.termsOfServiceCache.Len() != 0 {
+			if cacheItem, ok := s.rootStore.termsOfServiceCache.Get(s.rootStore.termsOfServiceCache.Keys()[len(s.rootStore.termsOfServiceCache.Keys())-1]); ok {
```

Comment:
you should be using `doStandardReadFromCache` here (that contains the get from cache and the metrics call)

Reply:
I don't love how we decide which is the "latest" here. I would suggest to add a new key there using for example "latest", and everytime we get the latest and is not in the cache, we add it to the cache.

Then in every save we remove the "latest" key of the cache.

Reply:
> I don't love how we decide which is the "latest" here. I would suggest to add a new key there using for example "latest", and everytime we get the latest and is not in the cache, we add it to the cache.
> 
> Then in every save we remove the "latest" key of the cache.

I definitely share some concerns here... its not pretty and a bit inefficient. But I'll admit I am not fully sure where you are suggesting I add the key (probably a product of being a newbee on Go).  Can I add a getLatest() method inside `mattermost-server/utils/lru.go` ?

Reply:
No, I'm thinking about adding a cache key for the latest, all the termsOfService keys are going to be 26 random characters, my proposal is something like `s.rootStore.doStandardReadCache(s.rootStore.termsOfServiceCache, "latest")`, and when you save a new termsOfService, you set the `"latest"` and the concrete cache key.

Reply:
@jespino please let me know if my latest changes are close to what you were thinking regarding the "latest" key on the termsOfServiceCache